File: KG.MES.UI.Shared/wwwroot/js/socketService.js
window.socketService = {
	socket: null,
	dotnetRef: null,

	connect: function (serverUrl) {
		console.log('');
		console.log('Socket connecting to:', serverUrl);
		console.log('');
		this.socket = io(serverUrl, {
			transports: ['websocket', 'polling']
		});

		this.socket.on('connect', () => {
			console.log('Socket connected');
		});

		this.socket.on('disconnect', () => {
			console.log('Socket disconnected');
		});
	},

	onMessage: function (dotnetRef) {
		this.dotnetRef = dotnetRef;
	},

	subscribe: function (channel, id) {
		if (this.socket) {
			var room = id ? channel + ':' + id : channel;
			console.log('Subscribing to:', room);
			this.socket.emit('subscribe:' + channel, id || '');

			this.socket.on(channel + ':updated', (data) => {
				console.log('Received ' + channel + ':updated:', data);
				if (this.dotnetRef) {
					this.dotnetRef.invokeMethodAsync('OnSocketMessage', channel, data);
				}
			});
		}
	},

	unsubscribe: function (channel) {
		if (this.socket) {
			this.socket.emit('unsubscribe:' + channel);
			this.socket.off(channel + ':updated');
		}
	},

	disconnect: function () {
		if (this.socket) {
			this.socket.disconnect();
			this.socket = null;
		}
	}
};

File: KG.MES.Shared/Services/SocketService.cs
using Microsoft.Extensions.Logging;
using KG.MES.Shared.Interfaces;
using Microsoft.Extensions.Configuration;
using Microsoft.JSInterop;

namespace KG.MES.Shared.Services
{
	public class SocketService : ISocketService
	{
		private readonly IJSRuntime _jsRuntime;
		private readonly ILogger<SocketService> _logger;
		private readonly string _serverUrl;
		private bool _isInitialized;

		public event Action<string, object>? OnMessage;

		public SocketService(IJSRuntime jsRuntime, IConfiguration configuration, ILogger<SocketService> logger)
		{
			_jsRuntime = jsRuntime;
			_logger = logger;
			_serverUrl = configuration["SocketIO:Url"] ?? "http://192.168.0.179:3000";
		}

		public async Task ConnectAsync()
		{
			await _jsRuntime.InvokeVoidAsync("socketService.connect", _serverUrl);
			await _jsRuntime.InvokeVoidAsync("socketService.onMessage", DotNetObjectReference.Create(this));
			_isInitialized = true;
		}

		[JSInvokable]
		public void OnSocketMessage(string channel, object data)
		{
			_logger.LogInformation("Socket message: {Channel}", channel);
			OnMessage?.Invoke(channel, data);
		}

		public async Task SubscribeAsync(string channel)
		{
			if (!_isInitialized) await ConnectAsync();
			await _jsRuntime.InvokeVoidAsync("socketService.subscribe", channel);
		}

		public async Task UnsubscribeAsync(string channel)
		{
			await _jsRuntime.InvokeVoidAsync("socketService.unsubscribe", channel);
		}

		public async Task DisconnectAsync()
		{
			await _jsRuntime.InvokeVoidAsync("socketService.disconnect");
			_isInitialized = false;
		}

		public async Task JoinOrderAsync(string productionOrderId)
		{
			await _jsRuntime.InvokeVoidAsync("socketService.joinOrder", productionOrderId);
		}

		public async Task SubscribeAsync(string channel, string? id = null)
		{
			await _jsRuntime.InvokeVoidAsync("socketService.subscribe", channel, id);
		}
	}
}